Add moderation replies using menu

llewelld opened the pull request:
Allows a reply to be included with each moderation which is visible in the history table on the moderation page, and to the experience author on the edit story page.

In this implementation, moderators can see all previous replies in the table, but the author can see only the most recent reply.

This is a rather minimal approach to implementing what's described in #487, but could be expanded on later. It doesn't cover the rules and example responses portion of the task, those will still need to be added in.

llewelld commented:
Converting to draft while I fix the failing tests.

llewelld commented:
The tests now pass, so I'm moving this out of draft again. The coverage isn't great, but that's due to new lines in `share_experience()` and it looks like it wouldn't make sense trying to address these until pull requet #472 (or its successor) has been merged in.

gedankenstuecke commented on static/css/main.css:
```diff
@@ -528,3 +528,34 @@ body {
+  opacity: 1;
+}
+
+/* COLLAPSABLE CELLS ON THE MODERATION PAGE */
```
This is a really nice touch! 👍

gedankenstuecke reviewed (approved):
This is great! Technically everything works as expected and I agree, that for the testing parts we can wait for the other PR with tests to be merged in at this point. 

There's a few front-end things we might want to improve, but not necessarily in this PR (also can use tomorrow's meet-up to gather feedback on this!).  

1. In the moderator view the two different reply types are not very obvious, maybe those could be made a bit more clear. 
2. On the `view my experience` page the reply does not show up (the one at `/main/view/<uuid>/`, while it shows up in the `edit` view)

Lastly, there's a potentially bigger piece of work, which would be to inform the user that their experience has been moderated when the status is set to approved/rejected. This could be done via email through the `oh_member.message()` function, which takes a subject & body as parameters and uses the OH API to request sending an email on our behalf through OH. But this could also be its own PR from my view!

gedankenstuecke commented:
I've just gone ahead and solved a merge conflict in our css! :)

gedankenstuecke reviewed (commented):
I left some comments/thoughts in the issue related to this pr (#487)

llewelld commented:
Rebased.

llewelld commented:
Rebased on top of #509.

PR #509 is now the "freetext" version of the changes. This version layers the moderation menu on top of it. I've updated the titles to reflect this.

gedankenstuecke commented:
Thanks so much also in this thread, @llewelld! Just converted this and its sibling PR to drafts as they are open for discussion :)

llewelld commented:
I've made some small changes (to make the menu options link to the code-of-conduct) in an additional commit in preparation for the monthly meeting tomorrow.

llewelld commented:
Rebased.

llewelld commented:
Rebased.

llewelld commented:
For info (and ease of referece) this requires migration. To view the migration list:
```
docker exec -it autspaces-web-1 python manage.py showmigrations
```
To apply the migration for this PR:
```
docker exec -it autspaces-web-1 python manage.py migrate main 0010_experiencehistory_change_reply
```
To remove the migration when changing back to main branch:
```
docker exec -it autspaces-web-1 python manage.py migrate main 0009_auto_20230327_1226
```

helendduncan commented on server/apps/main/templates/main/moderate_experience.html:
Minor to the point of being pedantic - but is there a situation where the title might also need to be moderatable_text?

gedankenstuecke reviewed (approved):
There's only some very minor comments from my side on this, as it works great! 

I've left a comment on where we could implement data validation to address #524. 

Otherwise I noticed that we're missing a high-level `<div class="container">` in the `moderate_experiences.html` that should go around the `<section id="share-form">` to avoid the whole viewport running over its boundaries :-)

llewelld commented on server/apps/main/helpers.py:
```diff
@@ -133,7 +133,7 @@ def get_oh_metadata(ohmember, uuid):
 
     if len(file) == 0:
-        file = None
+        file = [None]
```
This is a personal thing as I find it easier to manage resource lifetimes if functions have only a single entry and exit point (`with` alleviates this, but isn't always available). If you prefer the early return here I'd be happy to change it though?

gedankenstuecke replied:
No strong preference for either way, I was just genuinely curious in order to learn something myself, and that makes total sense 👍 🙂

llewelld commented on server/apps/main/helpers.py:
```diff
@@ -524,12 +531,43 @@ def filter_in_review(files):
+    # We catch exceptions in case the experience doesn't exist
+    # in the local database
+    try:
+        model = PublicExperience.objects.get(experience_id=experience_id)
```
Thanks, this is a good point, and I'm glad you flagged it. I've updated the variable name.

llewelld commented on server/apps/main/views.py:
```diff
@@ -198,10 +199,13 @@ def share_experience(request, uuid=False):
                 form = ShareExperienceForm(data)
                 title = "Edit experience"
                 moderation_status = data.get("moderation_status", "not reviewed")
+                change_reply, changed_at = get_latest_change_reply(uuid)
```
Nicely identified, thank you! I've added the replies to the invalid form path too.

gedankenstuecke replied:
great, thanks!

llewelld commented on server/apps/main/views.py:
```diff
@@ -457,6 +465,7 @@ def moderate_experience(request, uuid):
             data = {**unchanged_experience_details, **trigger_details}
 
             moderation_comments = data.pop("moderation_comments", None)
+            moderation_reply = data.pop("moderation_reply", "")
```
I've not made any changes here, hoping that this can be addressed by @mastoffel on #524.

llewelld commented:
> Otherwise I noticed that we're missing a high-level `<div class="container">` in the `moderate_experiences.html` that should go around the `<section id="share-form">` to avoid the whole viewport running over its boundaries :-)

Nicely spotted! I've wrapped the contents of the section with a container div.

Thanks everyone for the really helpful reviews. I've updated the code and will now squash the commits and migrations. I'll message back here once that's completed.

llewelld commented:
Commits squashed and rebased.

gedankenstuecke commented:
@llewelld Please feel free to push merge on this! 📌
